// rollup-plugin-pure@0.2.1/dist/index.cjs
'use strict';

const MagicString = require('magic-string');
const pluginutils = require('@rollup/pluginutils');
const stripLiteral = require('strip-literal');

function _interopDefaultCompat (e) { return e && typeof e === 'object' && 'default' in e ? e.default : e; }

const MagicString__default = /*#__PURE__*/_interopDefaultCompat(MagicString);

function PluginPure(options) {
  const FUNCTION_RE = new RegExp(
    `(?<!\\/\\* #__PURE__ \\*\\/ )(?<![a-zA-Z0-9_$])(${options.functions.join("|").replaceAll("$", "\\$")})\\s*\\(`,
    "g"
  );
  const FUNCTION_RE_SINGLE = new RegExp(
    `(?<!\\/\\* #__PURE__ \\*\\/ )(?<![a-zA-Z0-9_$])(${options.functions.join("|").replaceAll("$", "\\$")})\\s*\\(`
  );
  const filter = pluginutils.createFilter(options.include, options.exclude);
  return {
    name: "nuxt:pure-annotations",
    transform: {
      order: "post",
      handler(code, id) {
        if (!filter(id) || !FUNCTION_RE_SINGLE.test(code)) {
          return;
        }
        const s = new MagicString__default(code);
        const strippedCode = stripLiteral.stripLiteral(code);
        for (const match of strippedCode.matchAll(FUNCTION_RE)) {
          s.overwrite(match.index, match.index + match[0].length, "/* #__PURE__ */ " + match[0]);
        }
        if (s.hasChanged()) {
          return {
            code: s.toString(),
            map: options.sourcemap ? s.generateMap({ hires: true }) : void 0
          };
        }
      }
    }
  };
}

exports.PluginPure = PluginPure;
